### ISSUE TITLE
generate:entity-boilerplate clean up issues

### ISSUE THREAD

**Issue by rbaugh** [totten/civix#250]
I feel like at some point I read that the boilerplate was trying to limit the amount of functions written to files as they weren't always used. I thought that I had read where re-running the civix generate:entity-boilerplate was suppose to clean up unused functions but I can't seem to locate where that is now.

The issue I am having is that the `civix generate:entity-boilerplate` being re-run on an existing entity is removing the template generated functions from the entoty.civix.php file, but not also removing from the entity.php file. This is then resulting in a Fatal error as it is trying to call a function that no longer exists.

Example entity Name: training

#### Before re-running generate:entity-boilerplate
In training.php:
```
function training_civicrm_alterSettingsFolders(&$metaDataFolders = NULL) {
  _training_civix_civicrm_alterSettingsFolders($metaDataFolders);
}
```

In training.civix.php 
```
function _training_civix_civicrm_alterSettingsFolders(&$metaDataFolders = NULL) {
  $settingsDir = __DIR__ . DIRECTORY_SEPARATOR . 'settings';
  if (!in_array($settingsDir, $metaDataFolders) && is_dir($settingsDir)) {
    $metaDataFolders[] = $settingsDir;
  }
}
```

#### After re-running generate:entity-boilerplate
In training.php (function still exists)
```
function training_civicrm_alterSettingsFolders(&$metaDataFolders = NULL) {
  _training_civix_civicrm_alterSettingsFolders($metaDataFolders);
}
```

In training.civix.php _training_civix_civicrm_alterSettingsFolders is removed.

It seems that if it is going to clean up the entity.civix.php, it need to make sure to remove the same function from the entity.php. This is suppose to be an automated clean up, it should take care of both functions.

**Comment by MegaphoneJon** [comment]
@rbaugh I wonder if you're thinking about this blog post? https://civicrm.org/blog/totten/civix-v2205-how-remove-million-lines-extra-code

It sounds to me like `civix generate:entity-boilerplate` assumes your extension has been upgraded to the latest civix format. So on the one hand I think you could probably solve your problem with `civix upgrade` but there's also a bug you're encountering where perhaps `civix generate:entity-boilerplate` (and perhaps other civix actions) shouldn't fire without checking whether the extension is on the latest civix format.

**Comment by rbaugh** [comment]
@MegaphoneJon that is the article I had read and I did come back across it after I sent the issue in. I hadn't run the civix upgrade at the time in this one extension/entity but it is as if the 'civix generate:entity-boilerplate` is trying to clean things up as well since it did remove one function and not the other.

If it is going to remove anything, it should make sure nothing is calling it first so you don't have an error. The `civix upgrade` command does clean things up. Still seems like it is an issue though. Maybe it leaves the cleaning to the `civix upgrade` job or needs to have a check so that it isn't removing a function that is tied to another function.

**Comment by totten** [comment]
> (1) (@MegaphoneJon)  there's also a bug you're encountering where perhaps `civix generate:entity-boilerplate` (and perhaps other civix actions) shouldn't fire without checking whether the extension is on the latest civix format.

> (2) (@rbaugh) If it is going to remove anything, it should make sure nothing is calling it first so you don't have an error.

I think these two points are going in the same direction, but the first formulation is closer fit to how the code is organized.

With 22.05+, there is a general flag in `info.xml` (eg `<format>22.05.2</format>`) that can determine whether the code is up-to-date with its templates. This flag is the driver for `civix upgrade`. I think it would be straight-forward to add a guard on all `generate:*` commands, eg (pseudocode):

```php
function assertExtensionIsCurrent() {
  $actualVersion = $info->get()->xpath('civix/format');
  $expectedVersion = Services::upgradeList()->getHeadVersion();
  if (version_compare($actualVersion, $expectedVersion, '<=')) {
    throw new \Exception("This extension requires upgrades. Please run 'civix upgrade' before generating code.");
  }
}
```

If it had this guard, and if you ran `generate:entity-boilerplate` before doing the upgrade, then `generate:entity-boilerplate` would abort early -- it would not add or remove anything. (It would complain -- and then you'd have to run `civix upgrade`, which would be more conscientious.)

There could be other formulations -- eg `civix generate:entity-boilerplate` could go through a list of checks for specific functions/files. But (*as originally+currently written*) it really doesn't have the groundwork to support that. It just blindly regenerates several files from the current templates - regardless of specific changes, cross-references, side-effects, etc. (Tracking those fiddlier things is  what the `upgrade` command is for.)

**Comment by rbaugh** [comment]
@totten  I haven't created many extensions so using civix was new to me. While reading the docs for civix, it said any time you make a change to the schema you can re-run the `generate:entity-boilerplate`. So when the code was updated to fix the `orderTables` function, #237, I simply just re-ran the `generate:entity-boilerplate`. The thing was I had already updated civix so it did the partial code cleanup I mentioned here. 

I didn't know that I should likely use the `civix upgrade` first as it didn't really mention it in the docs. I pulled the latest release down and with the latest version, I ran the `civix upgrade` and then was able to run the `generate:entity-boilerplate` to test the `orderTables` fix.

**Comment by totten** [comment]
@rbaugh Thanks, yup, that makes perfect sense. The `upgrade` step is newer (less documented and less obvious). Hopefully, the new validator in #254 (released as v22.07.2) will make it smoother.
